homeassistant/components/reolink/host.py

```diff
@@ -0,0 +1,176 @@
+            enable_rtsp = True
+
+        if enable_onvif or enable_rtmp or enable_rtsp:
+            if not await self._api.set_net_port(
+                enable_onvif=enable_onvif,
+                enable_rtmp=enable_rtmp,
+                enable_rtsp=enable_rtsp,
+            ):
+                if enable_onvif:
+                    _LOGGER.error(
+                        "Unable to switch on ONVIF on %s. You need it to be ON to receive notifications",
+                        self._api.nvr_name,
+                    )
+
+                if enable_rtmp:
+                    _LOGGER.error(
+                        "Unable to switch on RTMP on %s. You need it to be ON",
+                        self._api.nvr_name,
+                    )
+                elif enable_rtsp:
+                    _LOGGER.error(
+                        "Unable to switch on RTSP on %s. You need it to be ON",
+                        self._api.nvr_name,
+                    )
+
+        if self._unique_id is None:
+            self._unique_id = format_mac(self._api.mac_address)
+
+        return True
+
+    async def update_states(self) -> bool:
+        """Call the API of the camera device to update the states."""
+        return await self._api.get_states()
+
+    async def disconnect(self):
+        """Disconnect from the API, so the connection will be released."""
+        try:
+            await self._api.unsubscribe_all()
+        except Exception as error:  # pylint: disable=broad-except
+            err = str(error)
+            if err:
+                _LOGGER.error(
+                    "Error while unsubscribing ONVIF events for %s: %s",
+                    self._api.nvr_name,
+                    err,
+                )
+            else:
+                _LOGGER.error(
+                    "Unknown error while unsubscribing ONVIF events for %s",
+                    self._api.nvr_name,
+                )
+
+        try:
+            await self._api.logout()
+        except Exception as error:  # pylint: disable=broad-except
+            err = str(error)
+            if err:
+                _LOGGER.error(
+                    "Error while logging out of %s: %s", self._api.nvr_name, err
+                )
+            else:
+                _LOGGER.error(
+                    "Unknown error while logging out of %s", self._api.nvr_name
+                )
+
+    async def stop(self, event=None):
+        """Disconnect the API."""
+        await self.disconnect()
+
+    def get_iohttp_session(self) -> aiohttp.ClientSession | None:
+        """Return the iohttp session."""
+        if self._clientsession is None or self._clientsession.closed:
+            context = ssl.create_default_context()
+            context.set_ciphers("DEFAULT")
+            context.check_hostname = False
+            context.verify_mode = ssl.CERT_NONE
+
+            self._clientsession = async_create_clientsession(
+                self._hass, verify_ssl=False
+            )
```

Comment:
Create a custom client session instead with its own ssl context by passing kwargs to `async_create_clientsession`

Reply:
The problem is that the ssl context needs to be passed as (ssl) option to the `aiohttp.TCPConnector`,
while the kwargs are passed as options to the `aiohttp.ClientSession`

https://github.com/home-assistant/core/blob/2c238baa24e76125a4c4175e93ec27c8704a8bc1/homeassistant/helpers/aiohttp_client.py#L108-L121

https://github.com/home-assistant/core/blob/2c238baa24e76125a4c4175e93ec27c8704a8bc1/homeassistant/helpers/aiohttp_client.py#L241-L258

Reply:
I could not figure out how to properly pass the ssl_context using the build in HomeAssistant funcitons.

When using the aiohttp functions directly I would do it like this:
```
SSL_CONTEXT=ssl.create_default_context()
SSL_CONTEXT.set_ciphers("DEFAULT")
SSL_CONTEXT.check_hostname = False
SSL_CONTEXT.verify_mode = ssl.CERT_NONE
aiohttp.ClientSession(timeout = self._timeout, connector = aiohttp.TCPConnector(ssl = SSL_CONTEXT))
```

Reply:
@bdraco could you please give some advise?
I have some time to work on it today and tommorow...

Reply:
I think you'll either need to make a PR to extend `async_create_clientsession` or not use it

Reply:
See the PR here: https://github.com/home-assistant/core/pull/84573
Not sure if this is the best way to do it, I am not so experianced with aiohttp, so I would like someone with knowladge to look over it.

For now, I have just removed the creation of the aiohttp session for this integration and let it be handled by the upstream library.